src/models/state/archival_state.rs

```diff
@@ -3561,4 +3582,56 @@ mod archival_state_tests {
             .await
             .unwrap();
     }
+
+    mod pow_witness {
+        use super::*;
+        use crate::tests::shared::fake_valid_sequence_of_blocks_for_tests;
+
+        #[traced_test]
+        #[tokio::test]
+        async fn stored_pow_witness_agrees_with_block_hash() {
+            let network = Network::Main;
+            let mut rng = thread_rng();
+            let mut archival_state = make_test_archival_state(network).await;
+            let genesis_block = Block::genesis_block(network);
+            let blocks: [Block; 3] = fake_valid_sequence_of_blocks_for_tests(
+                &genesis_block,
+                Timestamp::now(),
+                rng.gen(),
+            )
+            .await;
+            for block in blocks.iter() {
+                archival_state.write_block_as_tip(block).await.unwrap();
+            }
+
+            for block in blocks.iter() {
+                let block_digest = block.hash();
+                let stored_record = archival_state
+                    .block_index_db
+                    .get(BlockIndexKey::Block(block_digest))
+                    .await
+                    .unwrap()
+                    .as_block_record();
+                assert_eq!(
+                    block.hash(),
+                    stored_record
+                        .pow_witness
+                        .block_hash(stored_record.block_header),
+                    "Block hash from stored witness must agree with block hash for block height {}",
+                    block.header().height
+                );
```

Comment:
It took me a while to decipher this block of code. I think it is at best tangentially related to the purpose of the test as conveyed by the name. I would recommend moving it to separate test case or, if a test case with that purpose and function already exists, removing it altogether. If you insist on keeping it, please disclaim with comments that it is there for double-checking redundancy and not the main purpose of the test, so that later code readers do not stumble over it like I did.